litellm/llms/notdiamond.py

```diff
@@ -0,0 +1,253 @@
+        hash_content: Optional[bool] = False,
+    ) -> None:
+        locals_ = locals()
+        for key, value in locals_.items():
+            if key != "self" and value is not None:
+                setattr(self.__class__, key, value)
+    
+    @classmethod
+    def get_config(cls):
+        return {
+            k: v
+            for k, v in cls.__dict__.items()
+            if not k.startswith("__")
+            and not isinstance(
+                v,
+                (
+                    types.FunctionType,
+                    types.BuiltinFunctionType,
+                    classmethod,
+                    staticmethod,
+                ),
+            )
+            and v is not None
+            or k == "llm_providers"
+        }
+
+
+def validate_environment(api_key):
+    if api_key is None:
+        raise ValueError(
+            "Missing NotDiamond API Key - A call is being made to notdiamond but no key is set either in the environment variables or via params"
+        )
+    headers = {
+        "Authorization": "Bearer " + api_key,
+        "accept": "application/json",
+        "content-type": "application/json",
+        "User-Agent": f"litellm/{version}",
+    }
+    return headers
+
+
+def get_litellm_model(response: dict) -> str:
+    nd_provider = response['providers'][0]['provider']
+    nd_model = response['providers'][0]['model']
+    nd_provider_model = f"{nd_provider}/{nd_model}"
+    litellm_model = notdiamond2litellm[nd_provider_model]
+    return litellm_model
+
+
+def update_litellm_params(litellm_params: dict):
+    '''
+    Create a new litellm_params dict with non-default litellm_params from the original call, custom_llm_provider and api_base
+    '''
+    new_litellm_params = dict()
+    for k, v in litellm_params.items():
+        # all litellm_params have defaults of None or False, except force_timeout
+        if (k == "force_timeout" and v != 600) or v:
+            new_litellm_params[k] = v
+    if "custom_llm_provider" in new_litellm_params: del new_litellm_params["custom_llm_provider"]
+    if "api_base" in new_litellm_params: del new_litellm_params["api_base"]
+    if "api_key" in new_litellm_params: del new_litellm_params["api_key"]
+    return new_litellm_params
+
+
+def completion(
+    model: str,
+    messages: list,
+    api_base: str,
+    model_response: ModelResponse,
+    print_verbose: Callable,
+    encoding,
+    api_key,
+    logging_obj,
+    optional_params=None,
+    litellm_params=None,
+    logger_fn=None,
+):
+    headers = validate_environment(api_key)
+    completion_url = api_base
+
+    ## Load Config
+    config = litellm.NotDiamondConfig.get_config()
+    for k, v in config.items():
+        if (
+            k not in optional_params
+        ):
+            optional_params[k] = v
+
+    # separate ND optional params from litellm optional params
+    nd_params = ["llm_providers", "tools", "max_model_depth", "tradeoff", "preference_id", "hash_content"]
```

Comment:
If you move forward with setting `llm_providers` as required, you'll need to update this here.

Reply:
This is just a way to separate out different kinds of params. Namely, the ones for ND API call, recommended model's API call, and LiteLLM params.